File: projet-perso-1-java/LookAndSay.java
public class LookAndSay {

    static String nextStepSlow(String s) {
        String result = "";
        int i = 0;
        while (i < s.length()) {
            char c = s.charAt(i);
            int count = 0;
            while (i < s.length() && s.charAt(i) == c) { count++; i++; }
            result += count + "" + c;
        }
        return result;
    }

    static String nextStepFast(String s) {
        StringBuilder sb = new StringBuilder();
        int i = 0;
        while (i < s.length()) {
            char c = s.charAt(i);
            int count = 0;
            while (i < s.length() && s.charAt(i) == c) { count++; i++; }
            sb.append(count);
            sb.append(c);
        }
        return sb.toString();
    }

    public static void main(String[] args) {
        final int DEPTH = 40;
        System.out.println("=== Look and Say – comparaison String vs StringBuilder ===\n");

        long start = System.currentTimeMillis();
        String seq = "1";
        for (int i = 0; i < DEPTH; i++) seq = nextStepFast(seq);
        System.out.printf("[StringBuilder (rapide)] Profondeur %d | Longueur : %d | Temps : %d ms%n",
                DEPTH, seq.length(), System.currentTimeMillis() - start);

        start = System.currentTimeMillis();
        seq = "1";
        for (int i = 0; i < DEPTH; i++) seq = nextStepSlow(seq);
        System.out.printf("[String        (lente) ] Profondeur %d | Longueur : %d | Temps : %d ms%n",
                DEPTH, seq.length(), System.currentTimeMillis() - start);
    }
}

File: projet-perso-1/LookAndSay.cs
using System;
using System.Text;
using System.Diagnostics;

class LookAndSay
{
    // Version LENTE : string immuable → génère des milliers d'objets garbage
    static string NextStepSlow(string s)
    {
        string result = "";
        int i = 0;
        while (i < s.Length)
        {
            char c = s[i];
            int count = 0;
            while (i < s.Length && s[i] == c) { count++; i++; }
            result += count.ToString() + c; // ← nouvelle string à chaque fois !
        }
        return result;
    }

    // Version RAPIDE : StringBuilder → buffer mutable, zéro garbage intermédiaire
    static string NextStepFast(string s)
    {
        var sb = new StringBuilder();
        int i = 0;
        while (i < s.Length)
        {
            char c = s[i];
            int count = 0;
            while (i < s.Length && s[i] == c) { count++; i++; }
            sb.Append(count);
            sb.Append(c);
        }
        return sb.ToString();
    }

    static void Run(string label, Func<string, string> step, int depth)
    {
        var sw = Stopwatch.StartNew();
        string seq = "1";
        for (int i = 0; i < depth; i++)
            seq = step(seq);
        sw.Stop();
        Console.WriteLine($"[{label}] Profondeur {depth} | Longueur finale : {seq.Length} | Temps : {sw.ElapsedMilliseconds} ms");
    }

    static void Main()
    {
        const int DEPTH = 40;
        Console.WriteLine("=== Look and Say – comparaison string vs StringBuilder ===\n");
        Run("StringBuilder (rapide)", NextStepFast, DEPTH);
        Run("string       (lente) ", NextStepSlow, DEPTH);
    }
}
